crates/cheatcodes/tests/random.rs

```diff
@@ -0,0 +1,164 @@
+        uint256 min = 0;
+        uint256 max = 1000000000;
+        console.log(vm.randomUint(min, max));
+        console.log(vm.randomUint(min, max));
+        console.log(vm.randomUint(min, max));
+    }
+
+    function testDeterministicRandomAddress() public {
+        console.log(vm.randomAddress());
+        console.log(vm.randomAddress());
+        console.log(vm.randomAddress());
+    }
+}
+"#,
+    )
+    .unwrap();
+
+    // Extracts the test result section from the DeterministicRandomnessTest contract output.
+    fn extract_test_result(out: &str) -> &str {
+        let start = out
+            .find("for test/DeterministicRandomnessTest.t.sol:DeterministicRandomnessTest")
+            .unwrap();
+        let end = out.find("Suite result: ok.").unwrap();
+        &out[start..end]
+    }
+
+    // Run the test twice with the same seed and verify the outputs are the same.
+    let seed1 = "0xa1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5b6c7d8e9f0a1b2";
+    let out1 = cmd
+        .args(["test", "--fuzz-seed", seed1, "-vv"])
+        .assert_success()
+        .get_output()
+        .stdout_lossy();
+    let res1 = extract_test_result(&out1);
+
+    let out2 = cmd
+        .forge_fuse()
+        .args(["test", "--fuzz-seed", seed1, "-vv"])
+        .assert_success()
+        .get_output()
+        .stdout_lossy();
+    let res2 = extract_test_result(&out2);
+
+    assert_eq!(res1, res2);
+
+    // Run the test with another seed and verify the output differs.
+    let seed2 = "0xb1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5b6c7d8e9f0a1b2";
+    let out3 = cmd
+        .forge_fuse()
+        .args(["test", "--fuzz-seed", seed2, "-vv"])
+        .assert_success()
+        .get_output()
+        .stdout_lossy();
+    let res3 = extract_test_result(&out3);
+    assert_ne!(res3, res1);
+
+    // Run the test without a seed and verify the outputs differs once again.
+    cmd.forge_fuse();
+    let out4 = cmd.args(["test", "-vv"]).assert_success().get_output().stdout_lossy();
+    let res4 = extract_test_result(&out4);
+    assert_ne!(res4, res1);
+    assert_ne!(res4, res3);
+});
+
+// tests `freshStorage` cheatcode.
+forgetest_init!(test_fresh_storage, |prj, cmd| {
+    prj.wipe_contracts();
+    prj.insert_ds_test();
+    prj.insert_vm();
+    prj.clear();
+
+    prj.add_source(
+        "Counter.t.sol",
+        r#"pragma solidity 0.8.24;
+import {Vm} from "./Vm.sol";
+import {DSTest} from "./test.sol";
+contract Counter {
+    uint256 public a;
+    address public b;
+    int8 public c;
+    address[] public owners;
+
+    function setA(uint256 _a) public {
+        a = _a;
+    }
+
+    function setB(address _b) public {
+        b = _b;
+    }
+
+    function getOwner(uint256 pos) public view returns (address) {
+        return owners[pos];
+    }
+
+    function setOwner(uint256 pos, address owner) public {
+        owners[pos] = owner;
+    }
+}
+
+contract CounterFreshStorageTest is DSTest {
```

Comment:
These tests look good overall. I would also add several more tests:

- One that makes the storage fresh, then which reads/writes multiple storage values, and checks that we don't accidentally get aliasing between them somehow.
- One that makes the storage fresh, then reads two different values, and puts a precondition (assume) that they are distinct value (not equal to each other), and then does something with those values.
- One which reads a storage variable, but then bounds the result to something very small. This is to check if we can efficiently generate random numbers within a bound.

Reply:
sounds good, will add such

Reply:
added a test here that I *think* it covers first 2 bullets

https://github.com/foundry-rs/foundry/blob/46d23223687ea373500ac852e892645eefe2d997/crates/cheatcodes/tests/random.rs#L152-L165

not clear though on how a test for 3rd one should look like?

Reply:
Hmmm, the third test maybe doesn't make sense, I agree. That one can probably be left for now, and if we encounter unexpected behavior with the tool we can report it and get a test added then.

The main concern is that a user may want to do this:

```
vm.freshStorage(address(counter));
vm.bound(counter.a(), 10, 20);
```

This basically would be saying: set the storage to all symbolic, but the value of `a` should be between `10` and `20`. Currently, I assume this would cause a lot of reverts because the `counter.a()` read will happen before the call to `bound`.

I'm not sure there is anything you can do about this in Foundry, it should work fine in Kontrol. But instead, for Foundry, you would have to do:

```
vm.freshStorage(address(counter));
uint256 a = vm.freshUInt(256, 10, 20); // produces the fresh value _and_ bounds it
counter.setA(a);
```

Then we know that the storage is all symbolic, but the correct bounds exist on `a`.

Reply:
The above is psuedocode, btw, just trying to brainstorm how to solve this issue of having a fully symbolic storage, but also allowing some bounds on specific values.